### PR TITLE
Support auth migration using port alias

### PR DESCRIPTION
**Pull request by diemtvu**
**What this PR does / why we need it**:

This PR provides mechanism to create 'alias' port with different auth policy via service annotation. This is mainly used to upgrade Istio security (to mTLS) without downtime (assuming "envoy multiplex", i.e a feature that allows envoy to handle both mTLS and plaintext on the same port, is not available).

For more details, see doc

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

https://github.com/istio/istio/issues/1232

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

File: pilot/proxy/envoy/config.go [hunk 1]
```diff
@@ -707,13 +716,9 @@ func buildInboundListeners(mesh *proxyconfig.MeshConfig, sidecar proxy.Node,
...
 	// assumes that endpoint addresses/ports are unique in the instance set
 	// TODO: validate that duplicated endpoints for services can be handled (e.g. above assumption)
-	for _, instance := range instances {
-		endpoint := instance.Endpoint
-		servicePort := endpoint.ServicePort
-		protocol := servicePort.Protocol
-		cluster := buildInboundCluster(endpoint.Port, protocol, mesh.ConnectTimeout)
-		clusters = append(clusters, cluster)
-
+	buildListerner := func(cluster *Cluster, instanceServiceHostname string,
```
Comment (rshriram) [inline pilot/proxy/envoy/config.go]: nit: we don't follow this inline function thingy in the pilot code base. Can you please create appropriate function and call it explicitly?
also, typo. buildListener

File: pilot/proxy/envoy/config.go [hunk 2]
```diff
@@ -784,14 +789,25 @@ func buildInboundListeners(mesh *proxyconfig.MeshConfig, sidecar proxy.Node,
...
 		}
 	}
+	for _, instance := range instances {
+		endpoint := instance.Endpoint
+		servicePort := endpoint.ServicePort
+		protocol := servicePort.Protocol
+		cluster := buildInboundCluster(endpoint.Port, protocol, mesh.ConnectTimeout)
+		clusters = append(clusters, cluster)
+		buildListerner(cluster, instance.Service.Hostname, endpoint.Port, servicePort.AuthenticationPolicy, protocol, endpoint)
+		if servicePort.AuthMigrationPort != nil {
```
Comment (rshriram) [inline pilot/proxy/envoy/config.go]: its twice the work. can you use if-else and call build Listener only once?

File: pilot/proxy/envoy/discovery.go [hunk 3]
```diff
@@ -365,9 +365,14 @@ func (ds *DiscoveryService) ListAllEndpoints(request *restful.Request, response
 					if instance.AvailabilityZone != "" {
 						t = &tags{AZ: instance.AvailabilityZone}
 					}
+					hostPort := instance.Endpoint.Port
+					if port.AuthMigrationPort != nil &&
+						port.AuthMigrationPort.Active {
+						hostPort = port.AuthMigrationPort.Port
+					}
```
Comment (rshriram) [inline pilot/proxy/envoy/discovery.go]: why do you have to rewrite the discovery request? all you need is to supply a different set of endpoints for SDS requests.

Comment (rshriram) [reply pilot/proxy/envoy/discovery.go]: nm.. ignore my comment

File: pilot/proxy/envoy/discovery_test.go [hunk 4]
```diff
@@ -113,6 +113,7 @@ func compareResponse(body []byte, file string, t *testing.T) {
 	if err != nil {
 		t.Fatalf(err.Error())
 	}
+	ioutil.WriteFile("/tmp/"+file, body, 0644)
```
Comment (rshriram) [inline pilot/proxy/envoy/discovery_test.go]: ? is this leftover debugging stuff?

File: pilot/proxy/envoy/discovery_test.go [hunk 5]
```diff
@@ -629,6 +681,21 @@ func TestListenerDiscoverySidecarAuthOptOut(t *testing.T) {
...
+	// Auth migration for port 80 (http, index 0) and port 90 (tcp, index 1)
+	mock.HelloService.Ports[0].AuthMigrationPort = &model.AuthMigrationPort{16080, proxyconfig.AuthenticationPolicy_MUTUAL_TLS, false}
+	mock.HelloService.Ports[2].AuthMigrationPort = &model.AuthMigrationPort{16090, proxyconfig.AuthenticationPolicy_NONE, true}
+	ds := makeDiscoveryService(t, registry, &mesh)
+	url := fmt.Sprintf("/v1/listeners/%s/%s", "istio-proxy", mock.HelloProxyV0.ServiceNode())
+	response := makeDiscoveryRequest(ds, "GET", url, t)
+	compareResponse(response, "testdata/lds-v0-none-auth-migration.json", t)
+	mock.HelloService.Ports[0].AuthMigrationPort = nil
+	mock.HelloService.Ports[2].AuthMigrationPort = nil
+}
```
Comment (rshriram) [inline pilot/proxy/envoy/discovery_test.go]: why do you need to test listener discovery?

File: pilot/proxy/envoy/testdata/cds-ssl-context-migration.json.golden [hunk 6]
```diff
@@ -0,0 +1,192 @@
+{
+  "clusters": [
+   {
+    "name": "in.1081",
```
Comment (rshriram) [inline pilot/proxy/envoy/testdata/cds-ssl-context-migration.json.golden]: where is the alternate port for 16080?

File: pilot/proxy/envoy/testdata/lds-v0-none-auth-migration.json.golden [hunk 7]
```diff
@@ -0,0 +1,1002 @@
...
+   {
+    "address": "tcp://0.0.0.0:15001",
+    "name": "virtual",
+    "filters": [],
+    "bind_to_port": true,
+    "use_original_dst": true
+   },
+   {
+    "address": "tcp://0.0.0.0:443",
+    "name": "http_0.0.0.0_443",
```
Comment (rshriram) [inline pilot/proxy/envoy/testdata/lds-v0-none-auth-migration.json.golden]: I don't think this is right. We don't have any 443 listeners anymore. @ayj correct?

File: pilot/model/service.go [hunk 8]
```diff
@@ -64,6 +64,15 @@ type Service struct {
...
 }
 
+// AuthMigrationPort represents an alias port for a service port that would
+// be used for authentication migration.
+// As of speaking, this information is extracted from service annotations.
+type AuthMigrationPort struct {
+	Port                 int                              `json:"port,omitempty"`
+	AuthenticationPolicy proxyconfig.AuthenticationPolicy `json:"policy,omitempty"`
+	Active               bool                             `json:"active,omitempty"`
+}
```
Comment (rshriram) [inline pilot/model/service.go]: @diemtvu bit confused. if user is explicitly opening up a new port on the pod for auth migration, then isn't it much more simpler to use standard routing rules and point to the new port? Istio routing rules support port based destinations. 

```yaml
destination:
 name: foo
 port : 80
route:
  - destination:
        port: 80
     weight: 90
  - destination:
        port: 9080
     weight: 10
```

This feature can be used for other scenarios as well. This  is currently not implemented but supported in the spec. So, if you implement this, you can orchestrate auth migration easily. 

The solution you have has some issues: you are doing a 100% shift immediately. If something breaks, there is no way to fallback. so its not migration. Its a complete switch with no fallback, unless I change my service port definition.

I am also not seeing in the golden files, how the migration port (Specified in the service declaration with auth policy = true) manifests..

File: pilot/proxy/envoy/config.go (hunk 2 above)
Comment (diemtvu) [reply pilot/proxy/envoy/config.go]: The first buildListener is for the original port, which should always be created. The second call, already inside a condition, to add an extra listener for the migration port.

File: pilot/proxy/envoy/discovery_test.go (hunk 5 above)
Comment (diemtvu) [reply pilot/proxy/envoy/discovery_test.go]: We expect extra listeners are added for migraion ports (16080, 16090)

File: pilot/proxy/envoy/testdata/cds-ssl-context-migration.json.golden (hunk 6 above)
Comment (diemtvu) [reply pilot/proxy/envoy/testdata/cds-ssl-context-migration.json.golden]: Actually this golden file is no longer needed. I was implemented this in the way to add extra CDS (and extra LDS) for every migration port, but then change to add LDS only, and change SDS host when the active state flipped.

On the other hand, based on other comments, may be I should change it back to add CDS/SDS/LDS for migration port, and let the routing rules to shift traffic.

**Comment by rshriram** [comment]
I am getting more clarity here. Didn't you already add the facility to turn on or off auth per port? In such a case, is this PR even needed? In other words, I could update my service definition to add a new port annotated with auth-enable. I can test the new auth in my own time. Once satisfied, I can simply delete the non auth port.

The flow described above should work with your previous PR to enable per port auth, right? What additional functionality does this PR aim to provide?

**Comment by diemtvu** [comment]
Agree that this PR may not be needed.

On the other hand, there may be some 'limitation' (not sure if it's a bug or feature) for the extra port added via service definition. To make sure we are on the same page, here is the spec of my service:

`
spec:
  ports:
  - name: http
    port: 8000
  - name: http-diem
    port: 2810
    targetPort: 8000
`

From what I see, there is only one inbound listener created (for port 8000), and both outbound clusters are go to the same pod:port. As the result, we cannot have different auth policy for them. In order to support this, we probably need to change the SDS and LDS in a similar way as in this PR.

{
    "service-key": "httpbin.default.svc.cluster.local|http",
    "hosts": [
      {
        "ip_address": "10.44.2.14",
        "port": 8000
      }
    ]
  },
  {
    "service-key": "httpbin.default.svc.cluster.local|http-diem",
    "hosts": [
      {
        "ip_address": "10.44.2.14",
        "port": 8000
      }
    ]
  },

**Comment by diemtvu** [comment]
I redo this PR to simply add annotation to make (endpoint) alias port. Basically, user will use service definition to add a service port with (unique) pod port. Istio sidecar intercept traffic on this new port, and with annotation, it knows to forward traffic to the application on the alias port.

**Comment by rshriram** [comment]
Sorry for the delayed response. We don't even need the annotation. WE are adding the ability to route to a specific port, using route rule. So, we don't need to add any annotations that implicitly setup routing. Users can explicitly add routing rules to route to auth enabled port if they wish to do so. WDYT?

**Comment by diemtvu** [comment]
Do you mean  route rule can be used to set route traffic between receiver sidecar, say port 8080, to application port 80? If that the case, then it's probably sufficient for auth enable usecase.

**Comment by rshriram** [comment]
You don't have to do it at receiver sidecar. You can do it from the sender side itself. If you declare a new port for your service and set it to auth enabled, then pilot will create an auth enabled listener for that port. From the sender side, you write a route rule like

```from serviceX to serviceY:80, route to serviceY:8080```

**Pull request closed without merging**